device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/InternalClient.java

```diff
@@ -419,22 +420,28 @@ public ReportedPropertiesUpdateResponse updateReportedProperties(TwinCollection
 
             if (timedOut)
             {
-                throw new TimeoutException("Timed out waiting for service to acknowledge reported properties update");
+                throw new IotHubClientException(IotHubStatusCode.DEVICE_OPERATION_TIMED_OUT, "Timed out waiting for service to acknowledge reported properties update");
             }
         }
 
-        return new ReportedPropertiesUpdateResponse(statusCodeAtomicReference.get(), versionAtomicReference.get());
+        IotHubClientException exception = iotHubClientExceptionAtomicReference.get();
+        if (exception != null)
+        {
+            throw exception;
+        }
+
+        return responseAtomicReference.get();
     }
 
     /**
      * Get the twin for this client. This client must have subscribed to desired properties before this method can be called.
      *
      * @return The twin for this client
-     * @throws TimeoutException if the service fails to acknowledge the getTwin request within the default timeout.
      * @throws InterruptedException if the operation is interrupted while waiting on the getTwin request to be acknowledged by the service.
      * @throws IllegalStateException if this client is not open or if this client has not subscribed to desired properties yet.
+     * @throws IotHubClientException if the request is rejected by the service for any reason of if the synchronous operation times out.
      */
-    public Twin getTwin() throws InterruptedException, TimeoutException, IllegalStateException
+    public Twin getTwin() throws InterruptedException, IllegalStateException, IotHubClientException
```

Comment:
Where did you land on having multiple types of exceptions we throw, keeping the native types. I'm guessing based on this you've kept them, yes?

Reply:
I smooshed the TimeoutException into the IotHubClientException, but otherwise it is the same as we went over in the design meeting. IllegalStateException and InterruptedException are special enough in Java to warrant their own catch statement.